File: 128. Longest Consecutive Sequence/Main.java
import java.util.HashSet;
import java.util.Set;

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.longestConsecutive(new int[]{100, 4, 200, 1, 3, 2}));
    }
}

class Solution {
    public int longestConsecutive(int[] nums) {
        int longest = 0;

        Set<Integer> set = new HashSet<>(nums.length);

        for (int num : nums) {
            set.add(num);
        }

        for (int num : set) {
            if (!set.contains(num - 1)) {
                int count = 0;
                while (set.contains(num)) {
                    count++;
                    num++;
                }
                longest = Math.max(count, longest);
            }
            if (longest > nums.length / 2) break;
        }

        return longest;
    }
}

File: 128. Longest Consecutive Sequence/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(longestConsecutive([]int{100, 4, 200, 1, 3, 2}))
}

func longestConsecutive(nums []int) int {
	longest := 0
	m := make(map[int]bool)

	for _, num := range nums {
		m[num] = true
	}

	for key := range m {
		if _, ok := m[key-1]; !ok {
			count := 0

			ok := true
			for ok {
				key++
				_, ok = m[key]
				count++
			}
			longest = max(count, longest)
		}

		if longest > len(nums)/2 {
			break
		}
	}

	return longest
}
